src/stack-unix/tcpv4v6_socket.ml

```diff
@@ -78,6 +149,10 @@ let dst fd =
     in
     ip, port
 
+let unread fd buf =
+  let buf = Cstruct.append buf fd.buf in
+  fd.buf <- buf
```

Comment:
what needs to be handled (for a complete, general API) is if a lwt task is already in `Lwt_cstruct.read` -- where the read should be cancelled and the `buf` provided here being returned to the caller.